torch/csrc/dynamo/eval_frame.c

```diff
@@ -1161,29 +1208,106 @@ PyObject* torch_c_dynamo_eval_frame_init(void) {
   Py_INCREF(Py_None);
   eval_frame_callback_set(Py_None);
 
-  PyObject* module = PyModule_Create(&_module);
-  if (module == NULL) {
-    return NULL;
-  }
-
-#if IS_PYTHON_3_11_PLUS
+  #if IS_PYTHON_3_11_PLUS
   if (PyType_Ready(&THPPyInterpreterFrameType) < 0) {
-    return NULL;
+    goto failure;
   }
+  // TODO (kenjin) convert this to heap type. This operation will leak references on interpreter finalization IIRC.
   Py_INCREF(&THPPyInterpreterFrameType);
-  if (PyModule_AddObject(module, "_PyInterpreterFrame", (PyObject*)&THPPyInterpreterFrameType) != 0) {
-    return NULL;
+  if (PyModule_AddObject(mod, "_PyInterpreterFrame", (PyObject*)&THPPyInterpreterFrameType) < 0) {
+    // Need to decref on failure as PyModule_AddObject only steals a reference on success.
+    Py_DECREF(&THPPyInterpreterFrameType);
+    goto failure;
   }
 #endif
 
 
   if (PyType_Ready(&CacheEntryType) < 0) {
-    return NULL;
+    goto failure;
   }
+  // TODO (kenjin) convert this to heap type. This operation will leak references on interpreter finalization IIRC.
   Py_INCREF(&CacheEntryType);
-  if (PyModule_AddObject(module, "_CacheEntry", (PyObject *) &CacheEntryType) < 0) {
+  if (PyModule_AddObject(mod, "_CacheEntry", (PyObject *) &CacheEntryType) < 0) {
       Py_DECREF(&CacheEntryType);
-      return NULL;
+      goto failure;
+  }
+  return 0;
+
+failure:
+  PyThread_tss_delete(&eval_frame_callback_key);
+  Py_DECREF(Py_None);
+  return -1;
+}
+
+static PyModuleDef_Slot eval_frame_module_slots[] = {
+    {Py_mod_exec, eval_frame_mod_exec},
+    {0, NULL}
+};
+
+static int eval_frame_module_traverse(PyObject *mod, visitproc visit, void *arg) {
+    eval_frame_mod_state *st = (eval_frame_mod_state *)PyModule_GetState(mod);
+    CHECK(st != NULL);
+    Py_VISIT(st->guard_fail_hook);
+    Py_VISIT(st->guard_error_hook);
+    return 0;
+}
+
+static int eval_frame_module_clear(PyObject *mod) {
```

Comment:
how does this work with our reset?

Reply:
Our reset doesn't actually unitialize and unload the entire module -- it only clears the code cache entries and some other administrative cleanup. So this code isn't called. This is only relevant for the GC to break reference cycles that may possibly involve the module object itself.